### PR TITLE
Issue/30/debounce

### PR DESCRIPTION
**Pull request by raintonr**
This also resolves issues #31 & #47.

### CODE REVIEW

File: io-package.json [hunk 1]
```diff
@@ -146,10 +146,16 @@
...
+            "linux": [
+                "build-essential",
+                "python"
+            ]
         },
         "dependencies": [
-            {"js-controller":  ">=2.0.0"}
+            {
+                "js-controller": ">=2.0.0",
+                "admin": ">=3.0.0"
```
Comment (Apollon77) [inline io-package.json:157]: Please put admin in an own "globalDependencies" array. (dependencies are required on same host, globalDependencies are "system wide"

File: main.js [hunk 2]
```diff
@@ -346,48 +354,56 @@ function readValue(port) {
...
+                        read:  data.isInput,
+                        write: !data.isInput
                     },
                     native: {
                     },
                     type: 'state'
                 };
-                adapter.setObject('gpio.' + port + '.state', obj, () =>
-                    syncPortDirection(port, data, callback));
+                adapter.setObject('gpio.' + port + '.state', obj, () => {
```
Comment (Apollon77) [inline main.js:380]: When we do a bigger update we should replace setObject by extendObject

File: main.js [hunk 3]
```diff
@@ -346,48 +354,56 @@ function readValue(port) {
...
-                if (obj.common.read !== isInput) {
-                    obj.common.read  = isInput;
-                    obj.common.write = !isInput;
-                    adapter.setObject('gpio.' + port + '.state', obj, () =>
-                        syncPortDirection(port, data, callback));
-                } else {
+                if (obj.common.read !== data.isInput) {
+                    obj.common.read  = data.isInput;
+                    obj.common.write = !data.isInput;
+                    adapter.setObject('gpio.' + port + '.state', obj, () => {
```
Comment (Apollon77) [inline main.js:388]: here too

File: main.js [hunk 4]
```diff
@@ -426,19 +441,65 @@ function syncPortDirection(port, data, callback) {
...
+                            type:  'boolean',
+                            role:  'button',
+                            read:  false,
+                            write: true
+                        },
+                        native: {
+                        },
+                        type: 'state'
+                    };
+                    adapter.setObject(stateName, obj, callback);
```
Comment (Apollon77) [inline main.js:466]: here too

File: main.js [hunk 5]
```diff
@@ -518,6 +591,40 @@ function initPorts() {
...
+                // Ignore buttons as they are handled below
+                if (adapter.config.gpios[port].input == 'in') {
+                    adapter.log.debug('GPIO change on port ' + port + ': ' + value);
+                    if (debounceTimers[port] != null) {
+                        // Timer is running but state changed (must be back) so just cancel timer.
+                        clearTimeout(debounceTimers[port]);
+                        debounceTimers[port] = null;
+                    } else {
+                        // Start a timer and report to state if doesn't revert within given period.
+                        debounceTimers[port] = setTimeout((t_port, t_value) => {
```
Comment (Apollon77) [inline main.js:608]: please make sure to also clear all timeouts in "unload"

**Review by Apollon77 — CHANGES_REQUESTED** [review]
hey, great work, some comments.
YOu can also update adapter-core to 2.4.0 in deps, now that js.controller 2 is minimum dep

File: main.js (hunk 2 above)
Comment (raintonr) [reply main.js:380]: Well... I just followed the lead of what was already there for consistency, but OK, will update.

File: io-package.json (hunk 1 above)
Comment (raintonr) [reply io-package.json:161]: Done

File: main.js (hunk 5 above)
Comment (raintonr) [reply main.js:615]: Done

**Comment by Apollon77** [comment]
great, thank you! Can you write me 1-2 lines for changelog about the change?

**Comment by raintonr** [comment]
> great, thank you! Can you write me 1-2 lines for changelog about the change?

Sure... will do that in #49.

**Pull request merged**